### DISCUSSION TITLE
Kiali Graph failing to load

### DISCUSSION THREAD

**Discussion by nbjohnson** [kiali/kiali#5958] [Q&A] [answered]
I initially have my istio mesh set to permissive mode while migrating workloads into the istio mesh. During this time having the entire mesh set to PERMISSIVE I can view the mtls status of connections using the "Security" checkbox. My connection between Kiali and Prometheus shows that it negotiated mtls during permissive mode so I enabled mtls STRICT on the prometheus namespace. Kiali was still able to load the graph to show traffic. However, if I try to enable "Security" while viewing the prometheus namespace, which has mtls STRICT set, I get an "Error Loading Graph" error. But if I don't have the "Security" checkbox checked I can see the prometheus namespace activity just fine.

Cluster wide mtls: PERMISSIVE
Prometheus namespace mtls: STRICT

1. view kiali namespace (mtls permissive) with Security checkbox - graph loads fine and I can see the mtls security status of the connection kiali is making to prometheus to retrieve metrics for the graph
2. view kiali namespace (mtls permissive) and prometheus namespace  (mtls strict) without Security checkbox - graph loads fine but with no security status
3. view kiali namespace (mtls permissive) and prometheus namespace  (mtls strict) with Security checkbox - graph errors out

Error:
```
Error loading Graph
Cannot load the graph: Failed ID gen1: cluster=[unknown] namespace=[prometheus] workload=[unknown] app=[unknown] version=[latest] service=[] graphType=[versionedApp]
```

So is Kiali just unable to show the security status of a namespace with mtls STRICT? Or is this a data parsing error and Kiali should be able to show the mtls status of a namespace with mtls STRICT enabled? While migrating namespaces to the strict setting it would be nice if Kiali was able to show the mtls status without erroring out, otherwise I will have to continuously unselect namespaces that have already been set to strict during the migration

**Comment by jmazzitelli** [comment]
_converting to Q&A Discussion; not verified as a bug - if it is, we can create a bug issue later with the replication procedures_

Please tell us what versions of Kiali and Istio you are using, and how you installed Kiali.

**Reply by nbjohnson** [reply]
Istio: 1.17.1
Kiali: 1.63.2

I installed Kiali by applying a tweaked version of the Kiali sample provided in istio 1.17.1

**Reply by jmazzitelli** [reply]
When you see errors like that in the UI, grab the kiali pod logs and see what errors/messages are in there. Any error messages you see that look related?

**Reply by nbjohnson** [reply]
This is the error that pops up in the logs around the time that I try to load the security status of the mtls strict namespace. I dont see it show up in any of the other scenarios.
```
/opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
net/http.(*ServeMux).ServerHTTP(0xc005c401a?, {0x205d310, 0xc0029aa380}, 0xc00259ea00)
/opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2462 +0x149
net/http.serverHandler.ServeHTTP({0xc001c503f0?}, {0x205d310, 0xc0029aa380}, 0xc00259ea00)
/opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc000da46e0, {0x205e7d8, 0xc0019b0450})
/opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
/opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:3071 +0x4db
```

**Reply by jmazzitelli** [reply]
that doesn't provide much - looks like part of a stack trace of the http handling in server.go (which is the main entry point), but no root cause or panic message or function name indicating where in the code the request went. So we'd need to see the entire stack trace or log to figure out what is going on.

**Reply by nbjohnson** [reply]
Here are the additional debug logs:
```
2023-03-29T14:53:54Z ERR Failed ID gen1: cluster=[unknown] namespace=[prometheus] workload=[unknown] app=[unknown] version=[latest] service=[] graphType=[versionedApp]: goroutine 17102 [running]:
runtime/debug.Stack()
        /opt/hostedtoolcache/go/1.18.10/x64/src/runtime/debug/stack.go:24 +0x65
github.com/kiali/kiali/handlers.handlePanic({0x204fc00, 0xc000d67f50})
        /home/runner/work/kiali/kiali/handlers/graph.go:86 +0x214
panic({0x19c4ba0, 0xc00073b900})
        /opt/hostedtoolcache/go/1.18.10/x64/src/runtime/panic.go:838 +0x207
github.com/kiali/kiali/graph.Id({0xc0022e90f7?, 0xc001d60000?}, {0xc0022e9140?, 0x380?}, {0x0, 0x0}, {0xc0022e9140?, 0xc00158a148?}, {0xc0022e9120, 0x7}, ...)
        /home/runner/work/kiali/kiali/graph/types.go:206 +0xe34
github.com/kiali/kiali/graph/telemetry/istio/appender.SecurityPolicyAppender.addSecurityPolicy({{0xc000d06531, 0xc}, 0x1, 0xc0026653e0, 0x64245102, {{0x1d7cbb2, 0x8}, {0x1d7cbb2, 0x8}, {0x1d77746, ...}}}, ...)
        /home/runner/work/kiali/kiali/graph/telemetry/istio/appender/security_policy.go:267 +0xe5
github.com/kiali/kiali/graph/telemetry/istio/appender.SecurityPolicyAppender.populateSecurityPolicyMap({{0xc000d06531, 0xc}, 0x1, 0xc0026653e0, 0x64245102, {{0x1d7cbb2, 0x8}, {0x1d7cbb2, 0x8}, {0x1d77746, ...}}}, ...)
        /home/runner/work/kiali/kiali/graph/telemetry/istio/appender/security_policy.go:255 +0xc99
github.com/kiali/kiali/graph/telemetry/istio/appender.SecurityPolicyAppender.appendGraph({{0xc000d06531, 0xc}, 0x1, 0xc0026653e0, 0x64245102, {{0x1d7cbb2, 0x8}, {0x1d7cbb2, 0x8}, {0x1d77746, ...}}}, ...)
        /home/runner/work/kiali/kiali/graph/telemetry/istio/appender/security_policy.go:193 +0x1d10
github.com/kiali/kiali/graph/telemetry/istio/appender.SecurityPolicyAppender.AppendGraph({{0xc000d06531, 0xc}, 0x1, 0xc0026653e0, 0x64245102, {{0x1d7cbb2, 0x8}, {0x1d7cbb2, 0x8}, {0x1d77746, ...}}}, ...)
        /home/runner/work/kiali/kiali/graph/telemetry/istio/appender/security_policy.go:55 +0xde
github.com/kiali/kiali/graph/telemetry/istio.BuildNamespacesTrafficMap({_, _}, {0xc002665440, {0x0, {0xc001581ab0, 0x7, 0x7}}, 0x0, 0x1, 0xc0026653e0, ...}, ...)
        /home/runner/work/kiali/kiali/graph/telemetry/istio/istio.go:80 +0x903
github.com/kiali/kiali/graph/api.graphNamespacesIstio({_, _}, _, _, {{0x1d7db5c, 0x9}, {0x1d782fa, 0x5}, {{0xc000d06573, 0x15}, ...}, ...})
        /home/runner/work/kiali/kiali/graph/api/api.go:52 +0x126
github.com/kiali/kiali/graph/api.GraphNamespaces({_, _}, _, {{0x1d7db5c, 0x9}, {0x1d782fa, 0x5}, {{0xc000d06573, 0x15}, {0xdf8475800, ...}}, ...})
        /home/runner/work/kiali/kiali/graph/api/api.go:33 +0x2fb
github.com/kiali/kiali/handlers.GraphNamespaces({0x204fc00, 0xc000d67f50}, 0xc0012ca800)
        /home/runner/work/kiali/kiali/handlers/graph.go:51 +0x165
net/http.HandlerFunc.ServeHTTP(0x1d85894?, {0x204fc00?, 0xc000d67f50?}, 0x40d2e5?)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
github.com/kiali/kiali/routing.metricHandler.func1({0x205d460?, 0xc0021b8c40}, 0xc000687d01?)
        /home/runner/work/kiali/kiali/routing/router.go:145 +0x13b
net/http.HandlerFunc.ServeHTTP(0x205e7d8?, {0x205d460?, 0xc0021b8c40?}, 0xc000687d70?)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
github.com/kiali/kiali/handlers.AuthenticationHandler.Handle.func1({0x205d460, 0xc0021b8c40}, 0xc001955600)
        /home/runner/work/kiali/kiali/handlers/authentication.go:78 +0x53e
net/http.HandlerFunc.ServeHTTP(0x19bee20?, {0x205d460?, 0xc0021b8c40?}, 0xc00071fa40?)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
github.com/kiali/kiali/server.plainHttpMiddleware.func1({0x205d460?, 0xc0021b8c40?}, 0xc00206ab10?)
        /home/runner/work/kiali/kiali/server/server.go:152 +0x65
net/http.HandlerFunc.ServeHTTP(0xc001955500?, {0x205d460?, 0xc0021b8c40?}, 0x0?)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc000170600, {0x205d460, 0xc0021b8c40}, 0xc001955300)
        [/home/runner/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210](mailto:/home/runner/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210) +0x1cf
github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1({0x205d310, 0xc0029362a0}, 0x4c2c8ced18?)
        [/home/runner/go/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:336](mailto:/home/runner/go/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:336) +0x23c
net/http.HandlerFunc.ServeHTTP(0x7fe2c3e6f4e8?, {0x205d310?, 0xc0029362a0?}, 0x40d2e5?)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2084 +0x2f
net/http.(*ServeMux).ServeHTTP(0xc000d0651a?, {0x205d310, 0xc0029362a0}, 0xc001955300)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2462 +0x149
net/http.serverHandler.ServeHTTP({0xc00206a960?}, {0x205d310, 0xc0029362a0}, 0xc001955300)
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc000d62960, {0x205e7d8, 0xc003c380f0})
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
        /opt/hostedtoolcache/go/1.18.10/x64/src/net/http/server.go:3071 +0x4db
```

I also see some of these logs, but seems to be on more loads than just the one I am erroring on:
`2023-03-29T14:56:21Z DBG Skipping bad source telemetry [case 1] [prometheus] [unknown] [unknown]`

**Reply by jmazzitelli** [reply]
> Skipping bad source telemetry [case 1] [prometheus] [unknown] [unknown]

That means that the telemetry Istio is storing is either bad or unexpected by Kiali.

@jshaughn check out the stack trace in the security appender... didn't you harden some stuff recently to avoid panics? I wonder if this is one spot that needs to be hardened when telemetry is bad/unexpected?
EDIT: well, maybe this isn't a panic, but it is a error message logged that seems to indicate telemetry is bad

**Comment by jshaughn** [comment, marked as answer]
This panic is due to unexpected telemetry supplied by istio.  When activating the `Security` option Kiali performs additional queries to determine the mutual_tls status, it seems these queries are hitting something bad that is not being handled.    Because these issues have been on the rise, Kiali 1.65.0 adds more protection to help pass over invalid telemetry.  Please try 1.65 and see if it resolves your problem.

**Reply by nbjohnson** [reply]
So when running `1.65` in the kiali logs I am seeing `Skipping addSecurityPolicy` and `Skipping addPrincipal` with the Failed ID error that I had above. But the graph is no longer erroring out, I am able to browse the graph just fine even with the Security box checked while viewing a mtls strict namespace

**Reply by jshaughn** [reply]
That is exactly what you want to see.  We are still logging the fact that we hit bad/unexpected telemetry, but we are now ignoring it without a panic.  The bad telemetry is an Istio problem that is unlikely to be solved, as it is hard to predict and happens in various corner cases.  It doesn't seem to be enough of a problem to try and solve as it seems to be in addition to the good telemetry.
